pkg/ruler/rulespb/rules.proto

```diff
@@ -28,6 +28,8 @@ message RuleGroupDesc {
   // to the Prometheus Manager.
   repeated google.protobuf.Any options = 9;
   int64 limit =10;
+  google.protobuf.Duration queryOffset = 11
+      [(gogoproto.nullable) = false, (gogoproto.stdduration) = true];
```

Comment:
I'm trying to use this branch to set a global rule query offset, and that setting isn't being honored in my local setup. In my cortex config file, I'm using:

```yaml
limits:
  ruler_query_offset: 1m
```

I think the issue is that this `queryOffset` field needs to be nullable. Since it's non-nullable, it's being initialized as`Duration(0)` in `compat.go`, and the ruler only falls back to the global offset if the field is actually nil, which you can see here:

https://github.com/prometheus/prometheus/blob/b75e6353742e32c7e18997ebd6a75b77563e33b5/rules/group.go#L644-L654

This change seems to fix it:

```diff
diff --git a/pkg/ruler/ruler.go b/pkg/ruler/ruler.go
index d53f78506..8fec45acb 100644
--- a/pkg/ruler/ruler.go
+++ b/pkg/ruler/ruler.go
@@ -910,6 +910,10 @@ func (r *Ruler) getLocalRules(userID string, rulesRequest RulesRequest, includeB
 			}
 		}
 		interval := group.Interval()
+		var queryOffset *time.Duration
+		if offset := group.QueryOffset(); offset != 0 {
+			queryOffset = &offset
+		}
 
 		groupDesc := &GroupStateDesc{
 			Group: &rulespb.RuleGroupDesc{
@@ -918,7 +922,7 @@ func (r *Ruler) getLocalRules(userID string, rulesRequest RulesRequest, includeB
 				Interval:    interval,
 				User:        userID,
 				Limit:       int64(group.Limit()),
-				QueryOffset: group.QueryOffset(),
+				QueryOffset: queryOffset,
 			},
 
 			EvaluationTimestamp: group.GetLastEvaluation(),
diff --git a/pkg/ruler/rulespb/compat.go b/pkg/ruler/rulespb/compat.go
index 4f916d09c..7526062ad 100644
--- a/pkg/ruler/rulespb/compat.go
+++ b/pkg/ruler/rulespb/compat.go
@@ -13,9 +13,10 @@ import (
 
 // ToProto transforms a formatted prometheus rulegroup to a rule group protobuf
 func ToProto(user string, namespace string, rl rulefmt.RuleGroup) *RuleGroupDesc {
-	queryOffset := time.Duration(0)
+	var queryOffset *time.Duration
 	if rl.QueryOffset != nil {
-		queryOffset = time.Duration(*rl.QueryOffset)
+		offset := time.Duration(*rl.QueryOffset)
+		queryOffset = &offset
 	}
 	rg := RuleGroupDesc{
 		Name:        rl.Name,
@@ -48,13 +49,17 @@ func formattedRuleToProto(rls []rulefmt.RuleNode) []*RuleDesc {
 
 // FromProto generates a rulefmt RuleGroup
 func FromProto(rg *RuleGroupDesc) rulefmt.RuleGroup {
-	queryOffset := model.Duration(rg.QueryOffset)
+	var queryOffset *model.Duration
+	if rg.QueryOffset != nil {
+		offset := model.Duration(*rg.QueryOffset)
+		queryOffset = &offset
+	}
 	formattedRuleGroup := rulefmt.RuleGroup{
 		Name:        rg.GetName(),
 		Interval:    model.Duration(rg.Interval),
 		Rules:       make([]rulefmt.RuleNode, len(rg.GetRules())),
 		Limit:       int(rg.Limit),
-		QueryOffset: &queryOffset,
+		QueryOffset: queryOffset,
 	}
 
 	for i, rl := range rg.GetRules() {
diff --git a/pkg/ruler/rulespb/rules.proto b/pkg/ruler/rulespb/rules.proto
index 189c050eb..ed4e98a76 100644
--- a/pkg/ruler/rulespb/rules.proto
+++ b/pkg/ruler/rulespb/rules.proto
@@ -29,7 +29,7 @@ message RuleGroupDesc {
   repeated google.protobuf.Any options = 9;
   int64 limit =10;
   google.protobuf.Duration queryOffset = 11
-      [(gogoproto.nullable) = false, (gogoproto.stdduration) = true];
+      [(gogoproto.nullable) = true, (gogoproto.stdduration) = true];
 }
 
 // RuleDesc is a proto representation of a Prometheus Rule
```

Reply:
There's a related discussion of this in the original prometheus PR, here: https://github.com/prometheus/prometheus/pull/14061/files#r1617950193